### PR TITLE
Roll engine with rolled dart

### PR DESCRIPTION
**Pull request by aam**

### CODE REVIEW

**Comment by aam** [comment]
[AppVeyor builder](https://ci.appveyor.com/project/flutter/flutter/build/1.0.9406) is not happy: @devoncarew have you seen this?
```
⏩ RUNNING: cd .; bin\flutter.bat analyze --flutter-repo
Analyzing 1395 files...
[warning] Missing concrete implementations of 'Map.addEntries', 'Map.cast', 'Map.map', 'Map.removeWhere' and 3 more. (packages\flutter_tools\lib\src\vmservice.dart, line 1235, col 7)
[warning] Inconsistent declarations of 'update' are inherited from (Map<String, dynamic>) → void, (String, (dynamic) → dynamic, {ifAbsent: () → dynamic}) → dynamic. (packages\flutter_tools\lib\src\vmservice.dart, line 1235, col 7)
[error] Base class introduces an invalid override. The type of 'ServiceObject.update' ('(Map<String, dynamic>) → void') isn't a subtype of 'Map<String, dynamic>.update' ('(String, (dynamic) → dynamic, {ifAbsent: () → dynamic}) → dynamic'). (packages\flutter_tools\lib\src\vmservice.dart, line 1235, col 18)
[lint] 7 public members lack documentation (ran in 51.2s)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
ERROR: Last command exited with 1 (expected: zero).
```

**Comment by aam** [comment]
cc @jason-simmons

**Comment by jason-simmons** [comment]
Some changes to the Map interface landed today:
   https://github.com/dart-lang/sdk/commit/6f78471687bb099a818439add9324607e70dacfd

ServiceMap in flutter_tools/lib/src/vmservice.dart needs to be updated

**Comment by devoncarew** [comment]
I think @jason-simmons is right in that we'll need to update the flutter repo a bit in response to these changes (/cc @leafpetersen)

**Comment by aam** [comment]
After moving to dev.22 and fixing [few analysis issues](https://github.com/flutter/flutter/pull/14541#issuecomment-364027919) I'm seeing new ones:
```
╰─➤  bin/flutter analyze --flutter-repo                                                                                                                                              1 ↵
Analyzing 1401 files...
[error] Evaluation of this constant expression throws an exception. (packages/flutter/test/painting/decoration_test.dart, line 340, col 28)
[error] Evaluation of this constant expression throws an exception. (packages/flutter/test/painting/decoration_test.dart, line 341, col 28)
[error] Evaluation of this constant expression throws an exception. (packages/flutter/test/painting/decoration_test.dart, line 342, col 28)
[error] Evaluation of this constant expression throws an exception. (packages/flutter/test/painting/decoration_test.dart, line 339, col 28)
[error] Evaluation of this constant expression throws an exception. (packages/flutter/test/widgets/list_wheel_scroll_view_test.dart, line 17, col 9)
[error] Type parameters could not be inferred for the mixin 'AutomaticKeepAliveClientMixin' because the base class implements the mixin's supertype constraint 'State<StatefulWidget>' in multiple conflicting ways (packages/flutter/lib/src/widgets/dismissible.dart, line 192, col 83)
[lint] 7 public members lack documentation (ran in 45.8s)
```

**Comment by devoncarew** [comment]
/cc'ing @bwilkerson for the `Evaluation of this constant expression throws an exception` errors. It's likely that that test code will need to be updated.

**Comment by Hixie** [comment]
We're probably hitting https://github.com/dart-lang/sdk/issues/26980. It may be time to fix that bug.

**Comment by aam** [comment]
> In `_DismissibleState`, `AutomaticKeepAliveClientMixin` needs to be `AutomaticKeepAliveClientMixin<Dismissible>`

`AutomaticKeepAliveClientMixin` type parameter was removed in e122d5d1d5c6159c04877cbbcbffdf29cfdba017

**Comment by aam** [comment]
> AutomaticKeepAliveClientMixin type parameter was removed in e122d5d

@a-siva @mraleph any thoughts on whether we can revert https://github.com/flutter/flutter/pull/14544 to keep analyzer happy about mixin-inheritance type conflicts?

**Comment by aam** [comment]
Okay alternative approach is what 08f754e is doing which is to silence analyzer warnings

**Comment by aam** [comment]
Travis build passes but now flutter gallery run with `--preview-dart-2` fails
```
─➤  ../../bin/flutter run --preview-dart-2

Launching lib/main.dart on Android SDK built for x86 in debug mode...
Initializing gradle...                                1.9s
Resolving dependencies...                             2.2s
Running 'gradlew assembleDebug'...                   18.5s
Built build/app/outputs/apk/debug/app-debug.apk (36.7MB).
Installing build/app/outputs/apk/app.apk...           3.5s
I/FlutterActivityDelegate( 3298): onResume setting current activity to this
E/flutter ( 3298): [ERROR:topaz/lib/tonic/logging/dart_error.cc(16)] Unhandled exception:
E/flutter ( 3298): type '(ByteData) => Future' is not a subtype of type '(ByteData) => Future<ByteData>' of 'value'
E/flutter ( 3298): #0      ___HashVMBase&MapMixin&_LinkedHashMapMixin^^#T0&#T1^#T0&#T1.[]= (file:///b/build/slave/Linux_Engine/build/src/third_party/dart/runtime/lib/compact_hash.dart)
E/flutter ( 3298): #1      BinaryMessages.setMessageHandler (package:flutter/src/services/platform_messages.dart:103:16)
E/flutter ( 3298): #2      BasicMessageChannel.setMessageHandler (package:flutter/src/services/platform_channel.dart:64:22)
E/flutter ( 3298): #3      ___BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.initInstances (package:flutter/src/scheduler/binding.dart:200:30)
E/flutter ( 3298): #4      ____BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding.initInstances (package:flutter/src/painting/binding.dart:22:11)
E/flutter ( 3298): #5      _____BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding.initInstances (package:flutter/src/rendering/binding.dart:31:11)
E/flutter ( 3298): #6      ______BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding&WidgetsBinding.initInstances (package:flutter/src/widgets/binding.dart:241:11)
E/flutter ( 3298): #7      new BindingBase (package:flutter/src/foundation/binding.dart:51:5)
E/flutter ( 3298): #8      new _BindingBase&GestureBinding (package:flutter/src/widgets/binding.dart)
E/flutter ( 3298): #9      new __BindingBase&GestureBinding&ServicesBinding (package:flutter/src/widgets/binding.dart)
E/flutter ( 3298): #10     new ___BindingBase&GestureBinding&ServicesBinding&SchedulerBinding (package:flutter/src/widgets/binding.dart)
E/flutter ( 3298): #11     new ____BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding (package:flutter/src/widgets/binding.dart)
E/flutter ( 3298): #12     new _____BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding (package:flutter/src/widgets/binding.dart)
E/flutter ( 3298): #13     new ______BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding&WidgetsBinding (package:flutter/src/widgets/binding.dart)
E/flutter ( 3298): #14     new WidgetsFlutterBinding (package:flutter/src/widgets/binding.dart)
E/flutter ( 3298): #15     WidgetsFlutterBinding.ensureInitialized (package:flutter/src/widgets/binding.dart:906:11)
E/flutter ( 3298): #16     runApp (package:flutter/src/widgets/binding.dart:698:25)
E/flutter ( 3298): #17     main (file:///Users/aam/projects/flutter/test_root6/flutter/flutter/examples/flutter_gallery/lib/main.dart:13:3)
E/flutter ( 3298): #18     _startIsolate.<anonymous closure> (file:///b/build/slave/Linux_Engine/build/src/third_party/dart/runtime/lib/isolate_patch.dart:278:19)
E/flutter ( 3298): #19     _RawReceivePortImpl._handleMessage (file:///b/build/slave/Linux_Engine/build/src/third_party/dart/runtime/lib/isolate_patch.dart:164:12)
Syncing files to device Android SDK built for x86...  6.4s
```

**Comment by mraleph** [comment]
@aam this looks like a type inference bug, let dig into it.

**Comment by mraleph** [comment]
@aam I think you need to purge your build folder and delete any dill files left out. 

Something is broken in the rebuild logic. I have noticed it occasionally before - dill files are not properly rebuild on engine upgrades.

```console 
# In ~/src/flutter/flutter/examples/flutter_gallery
$ rm -rf build && find . -iname '*.dill' -exec rm '{}' ';'
```

This should fix it. However it might be worth looking into why dill is not regenerated for you.

File: packages/flutter/test/widgets/list_wheel_scroll_view_test.dart [hunk 1]
```diff
@@ -14,10 +14,10 @@ void main() {
   group('construction check', () {
     testWidgets('ListWheelScrollView needs positive diameter ratio', (WidgetTester tester) async {
       try {
-        new ListWheelScrollView(
+        new ListWheelScrollView( // ignore: prefer_const_constructors
```
[the remark this answers is not in the document]
Comment (aam) [reply packages/flutter/test/widgets/list_wheel_scroll_view_test.dart]: Ah, good to know. Done.

**Comment by aam** [comment]
> @aam I think you need to purge your build folder and delete any dill files left out.

You are right, that fixed the problem.

> Something is broken in the rebuild logic. I have noticed it occasionally before - dill files are not properly rebuild on engine upgrades.

I see. One thing that is missing is deps file generation when kernel file(rather than snapshot) is generated. Not sure if it could cause stale builds though.

**Review by cbracken — APPROVED** [review]

**Comment by Hixie** [comment]
This patch had a very regressive impact on a wide number of benchmarks, notably dart2 aot snapshot size,  and release size. It also caused the dart2 ios benchmarks to stop reporting.

**Comment by mraleph** [comment]
I think the regressions come from 83e0ca24fd59d9584a44abcddf90f1592a9ec4c0, not from this patch per se. We are investigating why iOS benchmark stopped reporting.

**Comment by Hixie** [comment]
Sorry, my bad. This patch had a minor regressive impact on the aot_snapshot size metrics. I just got confused as to which PR I was commenting on.

**Pull request merged**